rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2127 @@
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                // Insufficient data.
+                SCLogDebug!("Insufficient data while parsing DCERPC header");
+                -1
+            }
+            Err(nom::Err::Error(_)) => {
+                SCLogDebug!("EoF reached with the data");
+                -1
+            }
+            Err(_) => {
+                // Error, probably malformed data.
+                SCLogDebug!("An error occured while parsing DCERPC header");
+                -2
+            }
+        }
+    }
+
+    pub fn handle_bindctxitem(&mut self, input: &[u8], uuid_internal_id: u16) -> i32 {
+        let endianness = self.get_endianness();
+        match parser::parse_bindctx_item(input, endianness) {
+            Ok((leftover_bytes, ctxitem)) => {
+                let mut uuidentry = DCERPCUuidEntry::new();
+                uuidentry.uuid = ctxitem.uuid;
+                uuidentry.internal_id = uuid_internal_id;
+                uuidentry.ctxid = ctxitem.ctxid;
+                uuidentry.version = ctxitem.version;
+                uuidentry.versionminor = ctxitem.versionminor;
+                let pfcflags = self.get_hdr_pfcflags().unwrap_or(0);
+                // Store the first frag flag in the uuid as pfc_flags will
+                // be overwritten by new packets
+                if pfcflags & PFC_FIRST_FRAG > 0 {
+                    uuidentry.flags |= DCERPC_UUID_ENTRY_FLAG_FF;
+                }
+                if let Some(ref mut bind) = self.bind {
+                    SCLogDebug!("DCERPC BIND CtxItem: Pushing uuid: {:?}", uuidentry);
+                    bind.uuid_list.push(uuidentry);
+                }
+                (input.len() - leftover_bytes.len()) as i32
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                // Insufficient data.
+                SCLogDebug!("Insufficient data while parsing DCERPC BIND CTXItem");
+                -1
+            }
+            Err(_) => {
+                // Error, probably malformed data.
+                SCLogDebug!("An error occurred while parsing DCERPC BIND CTXItem");
+                -1
+            }
+        }
+    }
+
+    pub fn process_bind_pdu(&mut self, input: &[u8]) -> i32 {
+        let mut retval = 0;
+        let mut idx = 12; // Bytes consumed if parser returns OK would be 12
+        match parser::parse_dcerpc_bind(input) {
+            Ok((leftover_bytes, header)) => {
+                let numctxitems = header.numctxitems;
+                self.bind = Some(header);
+                for i in 0..numctxitems {
+                    retval = self.handle_bindctxitem(&input[idx as usize..], i as u16);
+                    if retval == -1 {
+                        return -1;
+                    }
+                    idx = retval + idx;
+                }
+                (input.len() - leftover_bytes.len()) as i32 + retval * numctxitems as i32
```

Comment:
I have trouble parsing this: can you explain what is being returned here?

Reply:
I'll add comments in the code. But,
Bytes parsed (DCERPCBind): `input.len() - leftover_bytes.len()`
Bytes parsed per bindctxitem: `retval * numctxitems` (`retval` would always be 44 if parser works fine.)
Total bytes parsed: Bytes parsed (DCERPCBind) + Bytes parsed per bindctxitem

Reply:
This use of retval is confusing, as it only takes the `retval` from the last loop iteration. So I'd suggest either updating a `consumed` variable from within the loop, or replacing the retval by a hardcoded value (44 in this case).